src/DayPicker.tsx

```diff
@@ -256,6 +263,118 @@ export function DayPicker(props: DayPickerProps) {
 
   const dataAttributes = getDataAttributes(props);
 
+  const rootRef = useRef<HTMLDivElement>(null);
+  const rootElementSnapshot = useRef<HTMLElement>(null);
+  const previousMonths = useRef(months);
+
+  useLayoutEffect(() => {
+    if (months.length === 0) {
+      return;
+    }
+
+    const isSameMonth = dateLib.isSameMonth(
+      months[0].date,
+      previousMonths.current[0].date
+    );
+    const slideLeft = dateLib.isAfter(
+      months[0].date,
+      previousMonths.current[0].date
+    );
+    previousMonths.current = months;
+    if (rootRef.current) {
+      const previousMonthElement =
+        rootElementSnapshot.current?.querySelector(`[data-month]`);
+      rootElementSnapshot.current = rootRef.current.cloneNode(
+        true
+      ) as HTMLElement;
+      if (isSameMonth) {
+        return;
+      }
+
+      const currentMonthElement = rootRef.current.querySelector(`[data-month]`);
+      if (
+        currentMonthElement &&
+        currentMonthElement instanceof HTMLElement &&
+        previousMonthElement &&
+        previousMonthElement instanceof HTMLElement
+      ) {
+        // animate new displayed month
+        currentMonthElement.style.overflow = "hidden";
+        const monthCaptionElement =
+          rootRef.current.querySelector(`[data-month-caption]`);
+        if (monthCaptionElement && monthCaptionElement instanceof HTMLElement) {
+          monthCaptionElement.style.animation = "fadeIn 0.3s forwards";
+        }
+
+        const weeksElement = rootRef.current.querySelector(`[data-weeks]`);
+        if (weeksElement && weeksElement instanceof HTMLElement) {
+          weeksElement.style.animation = slideLeft
+            ? "slideInRight 0.3s forwards"
+            : "slideInLeft 0.3s forwards";
+        }
+        // animate new displayed month end
+        const cleanUp = () => {
+          if (
+            monthCaptionElement &&
+            monthCaptionElement instanceof HTMLElement
+          ) {
+            monthCaptionElement.style.animation = "";
+          }
+          if (weeksElement && weeksElement instanceof HTMLElement) {
+            weeksElement.style.animation = "";
+          }
+          currentMonthElement.style.overflow = "";
+          if (currentMonthElement.contains(previousMonthElement)) {
+            currentMonthElement.removeChild(previousMonthElement);
+          }
+        };
+
+        // animate old displayed month
+        previousMonthElement.style.pointerEvents = "none";
+        previousMonthElement.style.position = "absolute";
+        previousMonthElement.style.overflow = "hidden";
+        previousMonthElement.setAttribute("aria-hidden", "true");
```

Comment:
Is `aria-hidden` added for safety, or it is actually required?

Reply:
Just to be safe, and it's technically correct, the previous month's DOM will be only mounted for the visual animation and should not represent anything semantically.

Reply:
Oh cool - I'd better skip this attribute if we are not sure it is needed. What do you think?

Reply:
I can remove it, but I still think it's more semantically correct having it because the old month being animated out of the page is just for visualization purposes, it should not have any semantic meaning to the page while it's animated.